nuxeo-features/nuxeo-platform-oauth/src/test/java/org/nuxeo/ecm/platform/oauth/tests/OAuth2ChallengeFixture.java

```diff
@@ -0,0 +1,351 @@
+        String redirect = cr.getHeaders().get("Location").get(0);
+        assertTrue(redirect.contains("error=" + OAuth2Error.UNAUTHORIZED_CLIENT.toString().toLowerCase()));
+    }
+
+    @Test
+    public void authorizationShouldValidateRedirectURI() {
+        // Invalid: no redirect URI
+        Map<String, String> params = new HashMap<>();
+        params.put(CLIENT_ID_PARAM, CLIENT_ID);
+        params.put(RESPONSE_TYPE_PARAM, CODE_RESPONSE_TYPE);
+        params.put(STATE_PARAM, STATE);
+
+        ClientResponse cr = responseFromGetAuthorizationWith(params);
+        assertEquals(400, cr.getStatus());
+
+        // Invalid: not starting with https
+        params.put(REDIRECT_URI_PARAM, "http://redirect.uri");
+
+        cr = responseFromGetAuthorizationWith(params);
+        assertEquals(302, cr.getStatus());
+        String redirect = cr.getHeaders().get("Location").get(0);
+        assertTrue(redirect.contains("error=" + OAuth2Error.INVALID_REQUEST.toString().toLowerCase()));
+
+        // Invalid: starting with http://localhost with localhost part of the domain name
+        params.put(REDIRECT_URI_PARAM, "http://localhost.somecompany.com");
+
+        cr = responseFromGetAuthorizationWith(params);
+        assertEquals(302, cr.getStatus());
+        redirect = cr.getHeaders().get("Location").get(0);
+        assertTrue(redirect.contains("error=" + OAuth2Error.INVALID_REQUEST.toString().toLowerCase()));
+
+        // Invalid: not matching the one from the registered client
+        params.put(REDIRECT_URI_PARAM, "https://unknown.uri");
+
+        cr = responseFromGetAuthorizationWith(params);
+        assertEquals(302, cr.getStatus());
+        redirect = cr.getHeaders().get("Location").get(0);
+        assertTrue(redirect.contains("error=" + OAuth2Error.INVALID_REQUEST.toString().toLowerCase()));
+
+        // Valid: not starting with http
+        registerClient("Nuxeo Mobile", "nuxeo-mobile-app", "", "nuxeo://authorize");
+        params.put(CLIENT_ID_PARAM, "nuxeo-mobile-app");
+        params.put(REDIRECT_URI_PARAM, "nuxeo://authorize");
+
+        cr = responseFromGetAuthorizationWith(params);
+        assertEquals(200, cr.getStatus());
+
+        // Valid: starting with http://localhost with localhost not part of the domain name
+        registerClient("Localhost", "localhost", "", "http://localhost:8080/nuxeo");
+        params.put(CLIENT_ID_PARAM, "localhost");
+        params.put(REDIRECT_URI_PARAM, "http://localhost:8080/nuxeo");
+
+        cr = responseFromGetAuthorizationWith(params);
+        assertEquals(200, cr.getStatus());
+
+        // Valid: starting with https
+        registerClient("Secure", "secure", "", REDIRECT_URI);
+        params.put(CLIENT_ID_PARAM, "secure");
+        params.put(REDIRECT_URI_PARAM, REDIRECT_URI);
+
+        cr = responseFromGetAuthorizationWith(params);
+        assertEquals(200, cr.getStatus());
+    }
+
+    @Test
+    public void shouldDenyAccess() {
```

Comment:
Maybe add a comment to say that the access is expected to be denied because of the missing `grant_acces` parameter